plugins/search-backend-module-catalog/src/alpha.ts

```diff
@@ -23,24 +23,24 @@ import {
   coreServices,
   createBackendModule,
 } from '@backstage/backend-plugin-api';
-import { TaskScheduleDefinition } from '@backstage/backend-tasks';
-import { searchIndexRegistryExtensionPoint } from '@backstage/plugin-search-backend-node/alpha';
-
+import { catalogServiceRef } from '@backstage/plugin-catalog-node/alpha';
 import {
+  CatalogCollatorEntityTransformer,
   DefaultCatalogCollatorFactory,
-  DefaultCatalogCollatorFactoryOptions,
 } from '@backstage/plugin-search-backend-module-catalog';
-import { catalogServiceRef } from '@backstage/plugin-catalog-node/alpha';
+import { searchIndexRegistryExtensionPoint } from '@backstage/plugin-search-backend-node/alpha';
+import { readScheduleConfigOptions } from './collators/config';
 
 /**
- * @alpha
  * Options for {@link searchModuleCatalogCollator}.
+ *
+ * @alpha
  */
-export type SearchModuleCatalogCollatorOptions = Omit<
-  DefaultCatalogCollatorFactoryOptions,
-  'discovery' | 'tokenManager' | 'catalogClient'
-> & {
-  schedule?: TaskScheduleDefinition;
+export type SearchModuleCatalogCollatorOptions = {
+  /**
+   * Allows you to customize how entities are shaped into documents.
+   */
+  entityTransformer?: CatalogCollatorEntityTransformer;
 };
```

Comment:
I'm wondering why you add this option here. If I understand the [documentation](https://backstage.io/docs/features/search/how-to-guides#how-to-customize-fields-in-the-software-catalog-index) correctly there is an alternative approach to `searchModuleCatalogCollator(...)` for adding the catalog collator. 

Did that changed in the new backend system? Is this approach easier?

Reply:
I'm looking at it from the angle that there's an old world and a new world, and both the old world and the new world need to continue functioning for now, in isolation from each other.

Without this option, you cannot use the new-world module export at all if you need to pass in a custom transformer. You'd instead be forced to copy paste the entire `createBackendModule` block and change how it's implemented slightly (passing in a new argument to `DefaultCatalogCollatorFactory`).

You'd have to change

```ts
backend.add(searchModuleCatalogCollator());
```

to

```ts
backend.add(
  createBackendModule({
    moduleId: 'catalogCollator',
    pluginId: 'search',
    register(env) {
      env.registerInit({
        deps: {
          config: coreServices.rootConfig,
          discovery: coreServices.discovery,
          tokenManager: coreServices.tokenManager,
          scheduler: coreServices.scheduler,
          indexRegistry: searchIndexRegistryExtensionPoint,
          catalog: catalogServiceRef,
        },
        async init({
          config,
          discovery,
          tokenManager,
          scheduler,
          indexRegistry,
          catalog,
        }) {
          indexRegistry.addCollator({
            schedule: scheduler.createScheduledTaskRunner(
              readScheduleConfigOptions(config),
            ),
            factory: DefaultCatalogCollatorFactory.fromConfig(config, {
              entityTransformer: // <--- NEW ARGUMENT GOES HERE
              discovery,
              tokenManager,
              catalogClient: catalog,
            }),
          });
        },
      });
    },
  }
)
```

and then you also realize that hmm `readScheduleConfigOptions` isn't exported, now you have to export that too etc :) Or refactor how the internals work a little, of course.

In general, we used to have the idea that the "escape hatch" when needing options that aren't static-config-friendly, you'd just reimplement the `createBackendModule`. But as you see, it easily turns into many MANY lines of code just to get the dependency injection in place, and it forces you to think hard about what escape-hatch-exports you need to do as well in addition to the plain module, to make this possible in the first place.

It's a tradeoff and both methods are viable. Still open for discussion really, and in all honesty this was just improvisation based on trying to keep the amount of effort to make a simple tweak low.

As a side note, this is also a reason that "extension points for modules" is a discussion. What if you could instead do

```ts
export createBackendSetup({
  deps: {
    collator: searchModuleCatalogCollatorExtenionPoint,
  },
  init({ collator }) {
    collator.setCustomEntityTransformer(entity => ... );
  }
});
```

and have that be auto discovered and injected into your backend? That's pseudocode of course, but hopefully gets the idea across.